Daemon / Reaper challenges

Currently, if I run:

```
flotter@pc:~/pebble/internals/daemon$ ./daemon.test -check.f TestRebootHelper -check.vv
START: daemon_test.go:751: daemonSuite.TestRebootHelper
START: daemon_test.go:61: daemonSuite.SetUpTest
PASS: daemon_test.go:61: daemonSuite.SetUpTest	0.001s

using shellcheck: "/usr/bin/shellcheck"
START: daemon_test.go:70: daemonSuite.TearDownTest
PASS: daemon_test.go:70: daemonSuite.TearDownTest	0.000s

... Panic: internal error: reaper must be started (PC=0x46F382)
```
This issue only happens when a subset of some tests are run.

The reason for this is because when the full suite is run, reapers started in some tests (e.g. ```TestAddRoutes```) **leak** into the following tests (this is a separate issue we need to clean up). For the daemon, the reaper is started when ```daemon.New(...)``` is called (indirectly when ```overlord.New``` create a new ```servstate``` manager). The reaper is killed when ```daemon.Stop(...)``` is called.

We have a much bigger issue here, which @benhoyt discovered, which is the current ```shutdown``` implementation in the daemon does not use the reaper, but calls ```exec.Command(...).CombinedOutput()``` directly. This means that the shutdown command will cause a ```wait()``` to reap the shutdown command on termination. This is in addition to the reaper, which will also receive a SIGCHILD and there will be a race who reaps the ```shutdown``` command.

Currently the ```shutdownHandler``` can be called while the reaper is active, and after the ```daemon.Stop()``` stopped the ```overlord``` where ```servstate.Stop()``` will kill the reaper. However the most common case today is likely as the final command in ```daemon.Stop()```, at which point the reaper will be dead.

One thing to note is that currently the reaper is owned by servstate. However, this is slightly misleading, because ```PR_SET_CHILD_SUBREAPER``` is a process level directive, so applies to the entire pebble, not only servstate. 

I have a proposal that I would like to get some feedback on:

1. Can we move the reaper right to the top so that the ```pebble run``` environment has this enabled for the entire process lifecycle?
2. Only use the reaper to run commands (this may require us to adapt ```systemd.go``` later on), which for now only means moving the ```shutdown``` call over to the reaper?
3. Have the daemon and overlord related test suites get the reaper enabled and disabled with each test, and only use ```FakeCommand``` with shellcheck running in reaper mode?

I think that the reaper is in all respects the better solution we have right now (preventing zombies from taking over our kernel process tables for long running setups) when running commands, and I see no downside from simply making this our preferred method to run and test with.

@benhoyt @barrettj12 @anpep Please let me know if you agree with the issue, and which way forward you would prefer ?

To further support this proposal:

Below is a test run where pebble received a SIGINT very quickly after startup. This is showing an interesting race between autostart and servstate shutdown, which kills the reaper. I need to fully investigate this, but moving the reaper out of servstate to a process level entity will also solve this.

flotter@pc:~/pebble$ 2023-09-28T12:02:11.993Z [pebble] Service "srv1" starting: sh -c  "while [ true ]; do echo .; sleep 1; done"
2023-09-28T12:02:11.996Z [pebble] POST /v1/services 6.401669ms 202
2023-09-28T12:02:11.996Z [pebble] Started default services with change 37.
2023-09-28T12:02:11.996Z [pebble] Exiting on terminated signal.
2023-09-28T12:02:12.007Z [pebble] Service "srv1" starting: sh -c  "while [ true ]; do echo .; sleep 1; done"
panic: internal error: reaper must be started

goroutine 66 [running]:
github.com/canonical/pebble/internals/reaper.StartCommand(0xc00057c160)
	/home/flotter/pebble/internals/reaper/reaper.go:164 +0x20f
github.com/canonical/pebble/internals/overlord/servstate.(*serviceData).startInternal(0xc000522480)
	/home/flotter/pebble/internals/overlord/servstate/handlers.go:435 +0xa66
github.com/canonical/pebble/internals/overlord/servstate.(*serviceData).start(0xc000522480)
	/home/flotter/pebble/internals/overlord/servstate/handlers.go:328 +0xc5
github.com/canonical/pebble/internals/overlord/servstate.(*ServiceManager).doStart(0xc0002060a0, 0xc000059df8?, 0xc000516000)
	/home/flotter/pebble/internals/overlord/servstate/handlers.go:133 +0x225
github.com/canonical/pebble/internals/overlord/state.(*TaskRunner).run.func1()
	/home/flotter/pebble/internals/overlord/state/taskrunner.go:195 +0xb4
gopkg.in/tomb%2ev2.(*Tomb).run(0xc000516000, 0x0?)
	/home/flotter/go/pkg/mod/gopkg.in/tomb.v2@v2.0.0-20161208151619-d5d1b5820637/tomb.go:163 +0x36
created by gopkg.in/tomb%2ev2.(*Tomb).Go
	/home/flotter/go/pkg/mod/gopkg.in/tomb.v2@v2.0.0-20161208151619-d5d1b5820637/tomb.go:159 +0xee
46ms

--- a/internals/cli/cmd_run.go
+++ b/internals/cli/cmd_run.go
@@ -29,6 +29,7 @@ import (
 	"github.com/canonical/pebble/cmd"
 	"github.com/canonical/pebble/internals/daemon"
 	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/reaper"
 	"github.com/canonical/pebble/internals/systemd"
 )
 
@@ -147,7 +148,18 @@ func sanityCheck() error {
 	return nil
 }
 
-func runDaemon(rcmd *cmdRun, ch chan os.Signal, ready chan<- func()) error {
+func runDaemon(rcmd *cmdRun, ch chan os.Signal, ready chan<- func()) (err error) {
+	err = reaper.Start()
+	if err != nil {
+		return fmt.Errorf("cannot start child process reaper: %w", err)
+	}
+	defer func() {
+		err = reaper.Stop()
+		if err != nil {
+			err = fmt.Errorf("cannot stop child process reaper: %w", err)
+		}
+	}()
+
 	t0 := time.Now().Truncate(time.Millisecond)
 
 	pebbleDir, socketPath := getEnvPaths()
@@ -157,7 +169,7 @@ func runDaemon(rcmd *cmdRun, ch chan os.Signal, ready chan<- func()) error {
 			return err
 		}
 	}
-	err := maybeCopyPebbleDir(pebbleDir, getCopySource())
+	err = maybeCopyPebbleDir(pebbleDir, getCopySource())
 	if err != nil {
 		return err
 	}

--- a/internals/daemon/api_exec_test.go
+++ b/internals/daemon/api_exec_test.go
@@ -32,6 +32,7 @@ import (
 	"github.com/canonical/pebble/client"
 	"github.com/canonical/pebble/internals/logger"
 	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/reaper"
 )
 
 var _ = Suite(&execSuite{})
@@ -46,6 +47,11 @@ func (s *execSuite) SetUpSuite(c *C) {
 }
 
 func (s *execSuite) SetUpTest(c *C) {
+	err := reaper.Start()
+	if err != nil {
+		c.Fatalf("cannot start reaper: %v", err)
+	}
+
 	socketPath := c.MkDir() + ".pebble.socket"
 	daemon, err := New(&Options{
 		Dir:        c.MkDir(),
@@ -64,6 +70,11 @@ func (s *execSuite) SetUpTest(c *C) {
 func (s *execSuite) TearDownTest(c *C) {
 	err := s.daemon.Stop(nil)
 	c.Check(err, IsNil)
+
+	err = reaper.Stop()
+	if err != nil {
+		c.Fatalf("cannot stop reaper: %v", err)
+	}
 }
 
 // Some of these tests use the Go client for simplicity.

--- a/internals/daemon/api_test.go
+++ b/internals/daemon/api_test.go
@@ -23,6 +23,7 @@ import (
 	"gopkg.in/check.v1"
 
 	"github.com/canonical/pebble/internals/overlord/restart"
+	"github.com/canonical/pebble/internals/reaper"
 )
 
 var _ = check.Suite(&apiSuite{})
@@ -38,6 +39,11 @@ type apiSuite struct {
 }
 
 func (s *apiSuite) SetUpTest(c *check.C) {
+	err := reaper.Start()
+	if err != nil {
+		c.Fatalf("cannot start reaper: %v", err)
+	}
+
 	s.restoreMuxVars = FakeMuxVars(s.muxVars)
 	s.pebbleDir = c.MkDir()
 }
@@ -46,6 +52,11 @@ func (s *apiSuite) TearDownTest(c *check.C) {
 	s.d = nil
 	s.pebbleDir = ""
 	s.restoreMuxVars()
+
+	err := reaper.Stop()
+	if err != nil {
+		c.Fatalf("cannot stop reaper: %v", err)
+	}
 }
 
 func (s *apiSuite) muxVars(*http.Request) map[string]string {

--- a/internals/daemon/daemon.go
+++ b/internals/daemon/daemon.go
@@ -43,6 +43,7 @@ import (
 	"github.com/canonical/pebble/internals/overlord/servstate"
 	"github.com/canonical/pebble/internals/overlord/standby"
 	"github.com/canonical/pebble/internals/overlord/state"
+	"github.com/canonical/pebble/internals/reaper"
 	"github.com/canonical/pebble/internals/systemd"
 )
 
@@ -680,7 +681,7 @@ func systemdModeReboot(rebootDelay time.Duration) error {
 	}
 	mins := int64(rebootDelay / time.Minute)
 	cmd := exec.Command("shutdown", "-r", fmt.Sprintf("+%d", mins), rebootMsg)
-	if out, err := cmd.CombinedOutput(); err != nil {
+	if out, err := reaper.CommandCombinedOutput(cmd); err != nil {
 		return osutil.OutputErr(out, err)
 	}
 	return nil

--- a/internals/daemon/daemon_test.go
+++ b/internals/daemon/daemon_test.go
@@ -39,6 +39,7 @@ import (
 	"github.com/canonical/pebble/internals/overlord/restart"
 	"github.com/canonical/pebble/internals/overlord/standby"
 	"github.com/canonical/pebble/internals/overlord/state"
+	"github.com/canonical/pebble/internals/reaper"
 	"github.com/canonical/pebble/internals/systemd"
 	"github.com/canonical/pebble/internals/testutil"
 )
@@ -60,6 +61,11 @@ type daemonSuite struct {
 var _ = Suite(&daemonSuite{})
 
 func (s *daemonSuite) SetUpTest(c *C) {
+	err := reaper.Start()
+	if err != nil {
+		c.Fatalf("cannot start reaper: %v", err)
+	}
+
 	s.pebbleDir = c.MkDir()
 	s.statePath = filepath.Join(s.pebbleDir, ".pebble.state")
 	systemdSdNotify = func(notif string) error {
@@ -73,6 +79,11 @@ func (s *daemonSuite) TearDownTest(c *C) {
 	s.notified = nil
 	s.authorized = false
 	s.err = nil
+
+	err := reaper.Stop()
+	if err != nil {
+		c.Fatalf("cannot stop reaper: %v", err)
+	}
 }
 
 func (s *daemonSuite) newDaemon(c *C) *Daemon {

--- a/internals/overlord/servstate/manager.go
+++ b/internals/overlord/servstate/manager.go
@@ -9,11 +9,9 @@ import (
 	"sync"
 	"time"
 
-	"github.com/canonical/pebble/internals/logger"
 	"github.com/canonical/pebble/internals/overlord/restart"
 	"github.com/canonical/pebble/internals/overlord/state"
 	"github.com/canonical/pebble/internals/plan"
-	"github.com/canonical/pebble/internals/reaper"
 	"github.com/canonical/pebble/internals/servicelog"
 )
 
@@ -71,11 +69,6 @@ func NewManager(s *state.State, runner *state.TaskRunner, pebbleDir string, serv
 		logMgr:        logMgr,
 	}
 
-	err := reaper.Start()
-	if err != nil {
-		return nil, err
-	}
-
 	runner.AddHandler("start", manager.doStart, nil)
 	runner.AddHandler("stop", manager.doStop, nil)
 
@@ -84,11 +77,6 @@ func NewManager(s *state.State, runner *state.TaskRunner, pebbleDir string, serv
 
 // Stop implements overlord.StateStopper and stops background functions.
 func (m *ServiceManager) Stop() {
-	err := reaper.Stop()
-	if err != nil {
-		logger.Noticef("Cannot stop child process reaper: %v", err)
-	}
-
 	// Close all the service ringbuffers
 	m.servicesLock.Lock()
 	defer m.servicesLock.Unlock()

--- a/internals/overlord/servstate/manager_test.go
+++ b/internals/overlord/servstate/manager_test.go
@@ -41,6 +41,7 @@ import (
 	"github.com/canonical/pebble/internals/overlord/servstate"
 	"github.com/canonical/pebble/internals/overlord/state"
 	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/reaper"
 	"github.com/canonical/pebble/internals/servicelog"
 	"github.com/canonical/pebble/internals/testutil"
 )
@@ -180,6 +181,11 @@ func (s *S) SetUpSuite(c *C) {
 }
 
 func (s *S) SetUpTest(c *C) {
+	err := reaper.Start()
+	if err != nil {
+		c.Fatalf("cannot start reaper: %v", err)
+	}
+
 	s.BaseTest.SetUpTest(c)
 
 	s.dir = c.MkDir()
@@ -210,6 +216,11 @@ func (s *S) TearDownTest(c *C) {
 	s.manager.Stop()
 	// General test cleanup
 	s.BaseTest.TearDownTest(c)
+
+	err := reaper.Stop()
+	if err != nil {
+		c.Fatalf("cannot stop reaper: %v", err)
+	}
 }
 
 func (s *S) TestDefaultServiceNames(c *C) {

--- a/internals/systemd/systemd.go
+++ b/internals/systemd/systemd.go
@@ -35,6 +35,7 @@ import (
 
 	"github.com/canonical/pebble/internals/osutil"
 	"github.com/canonical/pebble/internals/osutil/squashfs"
+	"github.com/canonical/pebble/internals/reaper"
 )
 
 var (
@@ -88,7 +89,8 @@ func (m *extMutex) Taken(errMsg string) {
 
 // systemctlCmd calls systemctl with the given args, returning its standard output (and wrapped error)
 var systemctlCmd = func(args ...string) ([]byte, error) {
-	bs, err := exec.Command("systemctl", args...).CombinedOutput()
+	cmd := exec.Command("systemctl", args...)
+	bs, err := reaper.CommandCombinedOutput(cmd)
 	if err != nil {
 		exitCode, _ := osutil.ExitCode(err)
 		return nil, &Error{cmd: args, exitCode: exitCode, msg: bs}
@@ -726,7 +728,8 @@ func (s *systemd) RemoveMountUnitFile(mountedDir string) error {
 		return err
 	}
 	if isMounted {
-		if output, err := exec.Command("umount", "-d", "-l", mountedDir).CombinedOutput(); err != nil {
+		cmd := exec.Command("umount", "-d", "-l", mountedDir)
+		if output, err := reaper.CommandCombinedOutput(cmd); err != nil {
 			return osutil.OutputErr(output, err)
 		}
 

